src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+            {
+                await FireTempPEDirtyAsync(fileName, true);
+            }
+        }
+
+        /// <summary>
+        /// Called privately to actually fire the TempPE events and optionally recompile the TempPE library for the specified inpu
+        /// </summary>
+        private async Task FireTempPEDirtyAsync(string fileName, bool shouldCompile)
+        {
+            // Not using use the ThreadingService property because unit tests
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            if (shouldCompile)
+            {
+                DesignTimeInputsItem inputs = AppliedValue.Value;
+                HashSet<string> files = GetFilesToCompile(fileName, inputs.SharedInputs);
+                string outputFileName = GetOutputFileName(inputs.OutputPath, fileName);
+                await CompileTempPEAsync(files, outputFileName);
+            }
+
+            _buildManager.OnDesignTimeOutputDirty(fileName);
+        }
+
+        public async Task<string> GetTempPEDescriptionXmlAsync(string fileName)
+        {
+            await InitializeAsync();
+
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+            if (fileName == null)
+                throw new ArgumentException("Must supply a file to build", nameof(fileName));
+
+            if (!inputs.Inputs.Contains(fileName))
+                throw new ArgumentException("FileName supplied must be one of the DesignTime source files", nameof(fileName));
+
+            string outputPath = inputs.OutputPath;
+
+            HashSet<string> files = GetFilesToCompile(fileName, inputs.SharedInputs);
+            string outputFileName = GetOutputFileName(outputPath, fileName);
+            if (CompilationNeeded(files, outputFileName))
+            {
+                // For parity with legacy we don't care about the compilation result: Legacy only errors here if it runs out of memory queuing the compilation
+                await CompileTempPEAsync(files, outputFileName);
+            }
+
+            // VSTypeResolutionService is the only consumer, and it only uses the codebase element so it's fine to default most of these (VC++ does the same)
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{fileName}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            DateTime outputDateTime = _fileSystem.LastFileWriteTime(outputFileName);
+
+            foreach (string file in files)
+            {
+                DateTime fileDateTime = _fileSystem.LastFileWriteTime(file);
```

Comment:
LastFileWriteTimeUtc

Reply:
What if this throw UnauthorizedAccessException/IOException?

Reply:
I tested with non-existent files and LastFileWriteTime doesn't throw, but its a good idea to add some belt and braces as I'm certainly not doing exhaustive testing around this.